src/main/java/me/jellysquid/mods/lithium/mixin/block/moving_block_shapes/PistonBlockEntityMixin.java

```diff
@@ -0,0 +1,129 @@
+
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.block.PistonBlock;
+import net.minecraft.block.PistonHeadBlock;
+import net.minecraft.block.entity.PistonBlockEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.util.shape.VoxelShape;
+import net.minecraft.util.shape.VoxelShapes;
+import net.minecraft.world.BlockView;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+import java.util.Collection;
+import java.util.WeakHashMap;
+
+
+/**
+ * @author 2No2Name
+ */
+@Mixin(PistonBlockEntity.class)
+public abstract class PistonBlockEntityMixin {
+    private static final VoxelShape[] PISTON_BASE_WITH_MOVING_HEAD_SHAPES = precomputePistonBaseWithMovingHeadShapes();
+    @SuppressWarnings("unchecked")
+    private static final ThreadLocal<WeakHashMap<VoxelShape, VoxelShape>[]> SHAPE_TO_OFFSET_SIMPLIFIED_SHAPE = ThreadLocal.withInitial(() -> new WeakHashMap[1 + (2 * 6)]);
+
+    @Shadow
+    private Direction facing;
+    @Shadow
+    private boolean extending;
+    @Shadow
+    private boolean source;
+    @Shadow
+    private float progress;
+
+    @Shadow
+    public abstract boolean isSource();
+
+    private static VoxelShape[] precomputePistonBaseWithMovingHeadShapes() {
+        //not all combinations of offset and shortMode are used in vanilla, but the rule is not clear
+        float[] offsets = {0f, 0.5f, 1f};
+        Direction[] directions = Direction.values();
+        Collection<Boolean> shortModes = PistonHeadBlock.SHORT.getValues();
+
+        VoxelShape[] mergedShapes = new VoxelShape[offsets.length * directions.length * shortModes.size()];
+
+        for (Direction facing : directions) {
+            VoxelShape baseShape = Blocks.PISTON.getDefaultState().with(PistonBlock.EXTENDED, true)
+                    .with(PistonBlock.FACING, facing).getCollisionShape(null, null);
+            for (boolean isShort : shortModes) {
+                VoxelShape headShape = (Blocks.PISTON_HEAD.getDefaultState().with(PistonHeadBlock.FACING, facing))
+                        .with(PistonHeadBlock.SHORT, isShort).getCollisionShape(null, null);
+                for (float offset : offsets) {
+                    VoxelShape offsetHead = headShape.offset(facing.getOffsetX() * offset,
+                            facing.getOffsetY() * offset,
+                            facing.getOffsetZ() * offset);
+                    mergedShapes[getIndex(isShort, facing.getId(), offset)] = VoxelShapes.union(baseShape, offsetHead);
+                }
+            }
+        }
+
+        for (VoxelShape shape : mergedShapes) {
```

Comment:
I think this iteration wastes CPU resources. Isn't it possible to move it after initialization to line 64?

Reply:
This iteration is just a sanity check. It would just throw an Exception if the getIndex function is somehow incorrect. As it is only run once per start of the game, I don't think it is important to avoid the loop, but instead of moving it, it can just be removed